Feature/02 post crud

tetetearraio916 opened the pull request:
## 作業内容
- postモデルとテーブルを作成
- アップロード機能にcarrierwaveを使用
- 画像リサイズ用にminimagickを使用
- 画像データ識別,複数投稿できるようにimageモデル、テーブルを作成
- userテーブルとpostテーブルの関連付け(1対多)
- imageテーブルとpostテーブルを関連付け(1対多)、同時にデータが保存されるよう設定
- post のCRUD処理の実装
- fakerを使用し、seedファイルにデータ追加
- 一度の複数画像投稿に対して、スライドできるようにswiperを使用
- 一度の複数画像投稿に対して、新規投稿ページのviewをjavascriptで動きが出るように実装( 元々の画像投稿フォームに変更をかなり加えてしまいました。)
　

tetetearraio916 committed:
postのモデル、テーブル、アップロード機能、swiperの実装

tetetearraio916 committed:
リサイズ用にminimagick導入

tetetearraio916 committed:
postモデルにimageカラム追加

tetetearraio916 committed:
Postのルーティング設定

tetetearraio916 committed:
post_controllerの設定

tetetearraio916 committed:
投稿、投稿一覧ページのview実装

tetetearraio916 committed:
postモデルとuserモデルの関連付けとアップロード画像のリサイズ機能の実装

tetetearraio916 committed:
各ユーザーごとの投稿一覧と画像リサイズの実装

tetetearraio916 committed:
編集、削除リンク実装

tetetearraio916 committed:
投稿詳細の実装とその他viewの修正

tetetearraio916 committed:
seedでのデータ追加

tetetearraio916 committed:
ja.yml削除

tetetearraio916 committed:
複数画像アップロードの修正とそれに伴うseedの変更

tetetearraio916 committed:
showのswiperの修正

DaichiSaito commented on app/assets/javascripts/application.js:
```diff
@@ -17,6 +17,8 @@
 //= require jquery/dist/jquery.js
 //= require popper.js/dist/umd/popper.js
 //= require bootstrap-material-design/dist/js/bootstrap-material-design.js
+//= input.js
+//= delete.js
```
命名が抽象的過ぎてわかりづらいのでもっと具体的に書いたほうが他の人にとって親切だと思います！

tetetearraio916 replied:
ファイル名をinput.jsとdelete.jsからpost_button_input.js、post_button_delete.jsに変更しました。コミットメッセージにコミットサブジェクトととしても記述しました。投稿した後の削除ボタンと勘違いしてしまう場合も想定して、どのような機能実装したファイルなのかメッセージにも詳細を書かせていただきました。

DaichiSaito commented on app/models/image.rb:
```diff
@@ -0,0 +1,22 @@
+#
+#  fk_rails_...  (post_id => posts.id)
+#
+class Image < ApplicationRecord
+  belongs_to :post
+  mount_uploader :url, ImageUploader
+end
```
carrierwaveのmultiple file uploads機能を使わなかった理由はなんでしょうか？個別に画像を消したり追加したりしたいからかな？

あとカラム名はurlじゃなくfileとかにしたほうが良いと思いますよ。ここにはURLが入るわけではないので。urlにしちゃうと`Image.find(1).url.url`とかって書くことになっちゃう。

tetetearraio916 replied:
最初はpostテーブルにimageカラムを作成してmultiple file uploade機能を実装してみたのですがうまくいかず。わからなかったため、ググってみるとこの[記事](https://qiita.com/a2cfry/items/4fa93a426eaf9967461a)を見つけ、他の方法で模索しようと思い別テーブルを作成して複数投稿できるよう実装しようと考えました。

カラム名に関しては、マイグレーションファイルをまとめる際に、Imageのurlのカラム名をfileにして、データ型をjson型からstring型に変更しました。

DaichiSaito commented on app/uploaders/image_uploader.rb:
```diff
@@ -0,0 +1,62 @@
+  # include CarrierWave::RMagick
+  include CarrierWave::MiniMagick
+
+  process resize_to_fill: [1000, 1000]
+
+
+  version :thumb do
+   process resize_to_fit: [223,223]
+  end
+
+  version :swiper do
+    process resize_to_fit: [400, 400]
+  end
+
+  version :thumbnil do
+    process resize_to_fit: [100, 100]
+  end
```
実務ではこれだけバージョン作る必要があるか要検討ですね！多ければ多いほどサーバ負荷かかるので。

あと`thumbnil`はタイポかな。

DaichiSaito commented on app/views/posts/index.html.slim:
```diff
@@ -0,0 +1,56 @@
+    .row
+      .col-md-8.col-12
+        - @posts.each do |post|
+            div class = "card mb-5 post" id = "post-#{post.id}"
```
`slim`だったら`.card.mb-5.post`のように書くのが一般的かなーと思います。

DaichiSaito commented on app/views/posts/index.html.slim:
```diff
@@ -0,0 +1,56 @@
+            div class = "card mb-5 post" id = "post-#{post.id}"
+              .card-header
+                .d-flex.align-items-center
+                  - @user=User.find(post.user_id)
```
N + 1問題がっつり発生するのでコントローラ側でイーガーロードしてこないと。

tetetearraio916 replied:
コントローラー側でPost.all.includes(:user)を使用し、view側でpost.usersにすることでN + 1問題に対して対応したつもりですが、対応策としては正しいでしょうか？

DaichiSaito commented on app/views/posts/show.html.slim:
```diff
@@ -0,0 +1,34 @@
+    .image-box
+      .swiper-container.top-carousel
+        .swiper-wrapper(style="transform: translate3d(0px, 0px, 0px)")
+          - @image = Image.where(post_id: @post.id)
```
こんな風にViewでインスタンス変数に何かセットするのは基本やめたほうがいいですよ。見通しが悪くなるので。

DaichiSaito replied:
あとやるなら変数名は`@images`にするべきですね。`where`は複数返すものなので。

このあとの`@image.each~~~`という書き方に違和感が出てしまう。

DaichiSaito commented on app/views/posts/show.html.slim:
```diff
@@ -0,0 +1,34 @@
+                i class = "far fa-edit fa-lg"
+      hr class = "m-0"
+
+= javascript_include_tag "https://unpkg.com/swiper/swiper-bundle.js"
```
これだけ外部から取ってくる理由は？

tetetearraio916 replied:
　javascriptのswiperが動かずはまっていたので、外部から取ってきていたのですが、nodeに導入したswiper-bundle.jsをapplication.js にrequireすることを忘れていたので修正し、= javascript_include_tag　"application"として変更しました。

DaichiSaito commented on app/views/users/index.html.slim:
```diff
@@ -0,0 +1,12 @@
+      .col-md-6.col-12.offset-md-3.mb-3
+        .card
+          .card-body
+            - @users.each do |user|
+              .user.mb-3.d-flex.justify-content-between
+                = link_to user_path(user) do
+                  = image_tag  "default.png", class: "rounded-circle mr-1"
+                  = user.name
```
こういう繰り返しはcollectionのパーシャルのrender機能を使ったほうが良いっすね。

DaichiSaito commented on db/migrate/20200919065303_add_image_to_posts.rb:
```diff
@@ -0,0 +1,5 @@
+class AddImageToPosts < ActiveRecord::Migration[5.2]
+  def change
+    add_column :posts, :image, :string
+  end
+end
```
マイグレーションファイル分け過ぎです。この他のも。

`db/migrate/20200919063311_create_posts.rb`の一つにまとまるように描きましょう！

tetetearraio916 replied:
user、post、imageのモデル事に分けてきれいにまとめました。

DaichiSaito commented on db/migrate/20200925111821_remove_user_from_image.rb:
```diff
@@ -0,0 +1,7 @@
+class RemoveUserFromImage < ActiveRecord::Migration[5.2]
+  def change
+    remove_foreign_key :images, :users
+    remove_index :images, :user_id
+    remove_reference :images, :user
+  end
+end
```
removeするなら最初から付けなければ良いですね。ゴミファイルがプルリクに上がってくるのは避けたいところ。

DaichiSaito reviewed (commented):
アップロードされた画像たちがプルリクにあがってきちゃってるので消しましょう。

tetetearraio916 committed:
change filename from input.js and delete.js to post_button_input.js and post_button_delete.js

ファイル名からどのような機能か推測できるよう変更しました
画像ファイルをアップロード時に追加で画像追加できるかつ削除できるjacascriptファイル　→　post_button_input.js
削除ボタンを押すとアップローしようとしている画像を投稿する前に削除できるjavascriptファイル　→　post_button_delete.js

tetetearraio916 committed:
organize migrationfile and change culumn from url to file

マイグレーションファイルがかなり作り込まれていたため、各モデル別にまとめました。

また、imageテーブルのurlのカラム名をfileに変更しました。
そのため、全てのview、javascriptファイルで使用されているurlをfileに変更する修正をさせていただきました。

tetetearraio916 committed:
Improvement　n+1 problem

n+1問題修正のためにコントローラ側のindexアクションにincludes (:users)を加えることで修正。
また、それぞれのwhereでデータを取得していたところをpost.imagesのように修正した。

tetetearraio916 committed:
Partialization of post form and list form

コードが見やすいように投稿フォームと投稿一覧をパーシャル化しました
